File: cvdl-rs/src/resume_layout.rs
use serde::{Deserialize, Serialize};

use crate::{layout_schema::Named, margin::Margin};

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub enum ColumnType {
    #[default]
    SingleColumn,
    DoubleColumn {
        vertical_margin: f32,
    },
}

impl ColumnType {
    pub fn vertical_margin(&self) -> f32 {
        match self {
            ColumnType::SingleColumn => 0.0,
            ColumnType::DoubleColumn { vertical_margin } => *vertical_margin,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct ResumeLayout {
    pub schema_name: String,
    pub column_type: ColumnType,
    pub margin: Margin,
    pub width: f32,
    pub height: f32,
}

impl Named for ResumeLayout {
    fn name(&self) -> &str {
        &self.schema_name
    }
}

impl ResumeLayout {
    pub fn from_json(json: &str) -> Vec<ResumeLayout> {
        let schemas: Vec<ResumeLayout> = serde_json::from_str(json).unwrap();
        schemas
    }
}

File: cvdl-ts/src/ResumeLayout.ts
import * as Margin from "./Margin";

export type ColumnType =
	| { tag: "SingleColumn" }
	| { tag: "DoubleColumn"; vertical_margin: number };

export const vertical_margin = (columnType: ColumnType) => {
	if (columnType.tag === "SingleColumn") {
		return 0;
	} else if (columnType.tag === "DoubleColumn") {
		return columnType.vertical_margin;
	}

	return 0;
};

export class ResumeLayout {
	schema_name: string;
	column_type: ColumnType;
	margin: Margin.t;
	width: number;
	height: number;

	constructor(
		schema_name: string,
		column_type: ColumnType,
		margin: Margin.t,
		width: number,
		height: number,
	) {
		this.schema_name = schema_name;
		this.column_type = column_type;
		this.margin = margin;
		this.width = width;
		this.height = height;
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	static fromJson(json: any): ResumeLayout {
		return new ResumeLayout(
			json.schema_name,
			json.column_type,
			json.margin,
			json.width,
			json.height,
		);
	}
}
